lib/compat.php

```diff
@@ -24,18 +24,90 @@
 	 * @return WP_Block_Type|false The registered block type on success, or false on failure.
 	 */
 	function register_block_type_from_metadata( $path, $args = array() ) {
-		$file = trailingslashit( $path ) . 'block.json';
-		if ( ! file_exists( $file ) ) {
+		$metadata_file = trailingslashit( $path ) . 'block.json';
+		if ( ! file_exists( $metadata_file ) ) {
 			return false;
 		}
 
-		$metadata = json_decode( file_get_contents( $file ), true );
-		if ( ! is_array( $metadata ) ) {
+		$metadata = json_decode( file_get_contents( $metadata_file ), true );
+		if ( ! is_array( $metadata ) || ! $metadata['name'] ) {
 			return false;
 		}
 
+		$block_dir           = dirname( $metadata_file );
+		$block_name          = $metadata['name'];
+		$asset_handle_prefix = str_replace( '/', '-', $block_name );
+
+		if ( isset( $metadata['editorScript'] ) ) {
+			$editor_script            = $metadata['editorScript'];
+			$editor_script_handle     = "$asset_handle_prefix-editor-script";
+			$editor_script_asset_path = realpath(
+				$block_dir . '/' . substr_replace( $editor_script, '.asset.php', -3 )
+			);
+			if ( ! file_exists( $editor_script_asset_path ) ) {
+				throw new Error(
+					"The asset file for the \"editorScript\" defined in \"$block_name\" block definition is missing."
+				);
```

Comment:
The choice to `throw` here is interesting. In general, it's a fine pattern to signal a failure in the registration. However, the function is already documented as returning "false on failure", which seems to be a conflicting approach.

Should this just be `return false;` ? 

How do we deal with a scenario where `editorScript` has been registered, and then a failure occurs in `script` or `style` registration? Undo those changes?

Reply:
I see now that `phpcs` doesn't like exceptions and the other issue you mentioned also is raised by the tool :)
I guess it should return `false` then.

I checked how the block registration is handled in WordPress core and I will update accordingly:

https://github.com/WordPress/wordpress-develop/blob/174ead229ddb6bd129347097bac6f998cde518b3/src/wp-includes/class-wp-block-type-registry.php#L57-L81

> How do we deal with a scenario where `editorScript` has been registered, and then a failure occurs in `script` or `style` registration? Undo those changes?

I haven't considered that. I guess the same issue applies as of today when you register those assets and the block registration fails during registration. Do you think that this scenario is likely when on production? I would expect it's helpful mostly when developing and thus this message that should help when debugging.

Reply:
A similar approach to what is in WordPress core proposed in: https://github.com/WordPress/gutenberg/pull/22519/commits/e6fce5e1c2c20f5a960035255d0244b827ed2589

Reply:
> Do you think that this scenario is likely when on production? I would expect it's helpful mostly when developing and thus this message that should help when debugging.

It's a good practice to strive for, but as you mention, it's likely not a common problem or one that a user would ever be faced with.